### ISSUE TITLE
vinyl: Tarantool crashes during transaction

### ISSUE THREAD

**Issue by aleksej-paschenko** [tarantool/tarantool#2487]
I use vinyl engine and tarantool version is 1.7.4-52-g980d300 from your ubuntu repository.
I have a transaction that changes 10000 items. 
When I press a ctrl-c to stop tarantool, it crashes 

```
#0  0x40ed41 in sig_fatal_cb(int)+129
#1  0x7fc9883e0390 in ?
#2  0x4ff630 in fiber_yield_timeout+96
#3  0x4e23c2 in ipc_cond_wait_timeout+66
#4  0x466740 in vy_range_tree_free_cb+384
#5  0x46701e in vy_index_delete+2222
#6  0x473d25 in vy_env_delete+85
#7  0x45e7a1 in VinylEngine::~VinylEngine()+33
#8  0x456650 in engine_shutdown()+48
#9  0x4c084e in box_free+30
#10 0x40ee52 in tarantool_free()+50
#11 0x7fc98757cff8 in ?
```

is this ok?

**Comment by kostja** [comment]
Please add a test case to reproduce this bug.

**Comment by aleksej-paschenko** [comment]
Minimal amount of code that can reproduce then bug:
```lua

box.cfg{
    listen = 3304,
    log = '2.log',
    log_level = 6,
    pid_file = '2.pid',
}

box.once('init', function ()

    -- Sale tuple: (account_id, id, date, cost)

    local sale_space = box.schema.create_space("sale", {engine = 'vinyl'})
    sale_space:create_index('primary',
        {type = 'TREE', parts={1, 'int', 2, 'int'}})

    box.schema.user.grant('guest', 'read,write,execute', 'universe')

    for i=1,1000000 do
        box.space.sale:insert{1, i, '2017-10-10', 0.0}
    end

end)

local msgpack = require('msgpack')

print ('press ctrl + c')
box.begin()
for _, t in box.space.sale.index.primary:pairs() do
    box.space.sale:replace{t[1], t[2], t[3], msgpack.NULL}
end
box.commit()
```
Please run tarantool app.lua several times and press ctrl + c after the corresponding message. 
I have a second stacktrace for this app.lua

#0  0x40ed41 in sig_fatal_cb(int)+129
#1  0x7f2ae616d390 in ?
#2  0x48500f in vy_mem_iterator_restore+1807
#3  0x464ab8 in vy_merge_iterator_next_key+984
#4  0x46e687 in vy_read_iterator_merge_next_key+55
#5  0x474211 in vy_read_iterator_next+113
#6  0x477ed6 in vy_cursor_next+278
#7  0x45faeb in iterator_next(iterator*)+27
#8  0x43ee4d in box_iterator_next+221
#9  0x4d48e8 in lbox_iterator_next+104
#10 0x50d577 in lj_BC_FUNCC+52
Aborted (core dumped)

**Comment by locker** [comment]
The first crash should be fixed by commit 053f8335b3f873734663e4ed25e98ad8a621b341
The second crash should be fixed by commit 3a73e5dcaa8c5e6cf647e94eb6dab98964c0193c
